### PR TITLE
Master survey owl yti

### PR DESCRIPTION
**Pull request by tivisse**
Description of the issue/feature this PR addresses:

Current behavior before PR:

Desired behavior after PR is merged:

--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

File: addons/survey/static/src/views/survey_views.js [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+import { KanbanRenderer } from "@web/views/kanban/kanban_renderer";
+import { listView } from '@web/views/list/list_view';
+import { kanbanView } from '@web/views/kanban/kanban_view';
+import { useService } from "@web/core/utils/hooks";
+import { patch } from "@web/core/utils/patch";
+
+const { useEffect, useRef } = owl;
+
+const SurveySampleMixin = {
+    useSurveyLoadSampleHook(selector) {
```
Comment (Williambraecky) [inline addons/survey/static/src/views/survey_views.js:14]: This function should be exported in a separate file and used in the `setup` of the component you wish to use it into.
Removing the need for patch.
See what we did for the time off to defer stuff [here](https://github.com/odoo/enterprise/pull/30468/files#diff-163974a0add5ddfb1f274aac4adf0930fb6c00946162a10c1f20a15791b54c5eR24)

File: addons/survey/static/src/views/survey_views.js [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+     * Load and show the sample survey related to the clicked element,
+     * when there is no survey to display.
+     * We currently have 3 different samples to load:
+     * - Sample Feedback Form
+     * - Sample Certification
+     * - Sample Live Presentation
+     */
+    async loadSample(method) {
+        // Prevent loading multiple samples if double clicked
+        const canCreate = this.props.archInfo.activeActions.create;
```
Comment (Williambraecky) [inline addons/survey/static/src/views/survey_views.js:56]: Actually this should never change, so we should call the hook only if create is true, and not check it in this function.

**Review by Williambraecky — CHANGES_REQUESTED** [review]
Here comes the smowl review

File: addons/survey/static/src/views/survey_views.js (hunk 1 above)
Comment (Williambraecky) [reply addons/survey/static/src/views/survey_views.js:14]: (Whether it is in a separate file or not does not matter btw)

File: addons/survey/static/src/views/survey_views.js [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+        useEffect(
+            (elems) => {
+                if (!elems || !elems.length) {
+                    return;
+                }
+                const handler = (ev) => {
+                    const surveyMethod = ev.currentTarget.closest('.o_survey_sample_container').getAttribute('action');
+                    this.loadSample(surveyMethod);
+                }
+                for (const elem of [...elems]) {
```
Comment (Williambraecky) [inline addons/survey/static/src/views/survey_views.js:25]: Actually I was wrong, we don't need to unpack elems into a list.
Same in the cleanup function
```suggestion
                for (const elem of elems) {
```

**Comment by tivisse** [comment]
Thanks for the review @Williambraecky :)

File: addons/survey/static/src/views/survey_views.js [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+
+const { useComponent, useEffect, useRef } = owl;
+
+export function useSurveyLoadSampleHook(selector) {
+    const component = useComponent();
+    component.rootRef = useRef("root");
+    component.actionService = useService("action");
+    component.orm = useService('orm');
+    component.isLoadingSample = false;
+    useEffect(
```
Comment (awa-odoo) [inline addons/survey/static/src/views/survey_views.js:18]: I don't know owl so this part confuses me greatly ^^

Not sure what useEffect does exactly, does it take a function as 1st argument and the arguments to pass to it as 2nd argument? It looks very strange.
And the part where you return another function that removes the eventHandler... -> 🤯 

Probably a stupid suggestion but how about a simple "onMounted", that way we don't have to worry about removing the handlers?

```js
onMounted(
    () => {
        const sampleTypes = component.rootRef.el ? component.rootRef.el.querySelectorAll(selector) : [];
        for (const sampleType of sampleTypes) {
            // (would let 'loadSample' extract the method to make this more simple)
            sampleType.addEventListener('click', component.loadSample);
        }
    }
)
```

**Review by awa-odoo — COMMENTED** [review]
Hello @tivisse 

Code is looking good, I added a suggestion for that useEffect thing ^^

Thanks ! 🐗

File: addons/survey/static/src/views/survey_views.js [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+    );
+    /**
+     * Load and show the sample survey related to the clicked element,
+     * when there is no survey to display.
+     * We currently have 3 different samples to load:
+     * - Sample Feedback Form
+     * - Sample Certification
+     * - Sample Live Presentation
+     */
+    component.loadSample = async function (method) {
```
Comment (Williambraecky) [inline addons/survey/static/src/views/survey_views.js:46]: I wouldn't store a function on the component like that, as it's not meant to be called by anything else anyway

File: addons/survey/static/src/views/survey_views.js (hunk 4 above)
Comment (Williambraecky) [reply addons/survey/static/src/views/survey_views.js:18]: We could have used onMounted, but we would have had to add onPatched aswell doing the same thing.
Also doing `component.loadSample` means we have to implement the function on the component each time we want to use the hook, which isn't great IMO


`useEffect` takes two functions, the first one is the effect which should return the cleanup, the second one is the dependencies.
`useEffect` is really well explained [here](https://github.com/odoo/owl/blob/master/doc/reference/hooks.md#useeffect)

**Comment by tivisse** [comment]
@robodoo r+

**Pull request closed without merging**